test/testautovariables.cpp

```diff
@@ -1745,7 +1746,7 @@ class TestAutoVariables : public TestFixture {
               "    const auto& str_cref2 = g(std::string(\"hello\"));\n"
               "    std::cout << str_cref2 << std::endl;\n"
               "}\n");
-        ASSERT_EQUALS("error", errout.str());
+        ASSERT_EQUALS("[test.cpp:5] -> [test.cpp:1] -> [test.cpp:2] -> [test.cpp:5] -> [test.cpp:6]: (error) Using reference to dangling temporary.\n", errout.str());
```

Comment:
@pfultz2  Interesting "side-effect"!

Reply:
hmm.. are you 100% sure there is a bug.. I just know that "The C++ standard guarantees that binding a temporary to a const reference on the stack, extends the lifetime of the temporary to the lifetime of the const reference. "

Reply:
> Interesting "side-effect"!

Actually, its nto a side effect. This PR doesn't change the error message, it just enables these tests to run.

> are you 100% sure there is a bug

Yes, C++ will extend the lifetime for temporaries when used directly, but the compiler doesn't when going through functions indirectly. The compiler will end the lifetime to `std::string("hello")` at the end of the function call. It doesn't know that `g` returned a reference to a stack variable. See [here](https://abseil.io/tips/107):

> This doesn’t work when you’re getting the subobject indirectly: the compiler doesn’t look through function calls (getters or the like).

And see [here](https://stackoverflow.com/a/2784304/375343):

> The lifetime extension is not transitive through a function argument. §12.2/5 [class.temporary]:
> > The second context is when a reference is bound to a temporary. The temporary to which the reference is bound or the temporary that is the complete object to a subobject of which the temporary is bound persists for the lifetime of the reference except as specified below. A temporary bound to a reference member in a constructor’s ctor-initializer (§12.6.2 [class.base.init]) persists until the constructor exits. A temporary bound to a reference parameter in a function call (§5.2.2 [expr.call]) persists until the completion of the full expression containing the call.

Reply:
BTW, a run using that exampe code with clang/asan also triggers a warning.